### PR TITLE
[ML] Natural Language Processing tasks and models

### PR DESCRIPTION
**Pull request by davidkyle**
Following on from #72218 which defined how large PyTorch models can be stored, this PR introduces the concepts of Natural Language Processing _tasks_ and defines a way to evaluate BERT models.

Mask Fill and Named Entity Recognition tasks are implemented here but others could be easily added now the framework is in place. In particular this PR implements tokenisation of input text for BERT models and defines a structure for post-graph processing. 

Once the PyTorch model is uploaded a trained model config referencing it must be PUT
```
PUT ml/trained_models/bert-model-for-maskfill
{
    "description": "Mask fill model",
    "model_type": "pytorch",
    "inference_config": {
        "classification": {
            "num_top_classes": 1
        }
    },
    "input": {
        "field_names": ["text_field"]
    },
    "location": {
        "index": {
            "model_id": "bert-model-for-maskfill",
            "name": "big_model"
        }
    }
}
```

And the model deployed:
```
POST _ml/trained_models/deployment/bert-model-for-maskfill/_start
```

### Mask Fill Example
```
POST _ml/trained_models/deployment/bert-model-for-maskfill/_infer
{
  "input": "Paris is the [MASK] of France."
}
```

Returns 
```
[
  {
    "token" : "capital",
    "score" : 0.9861745037766138,
    "sequence" : "Paris is the capital of France."
  },
  {
    "token" : "center",
    "score" : 0.00372138405614492,
    "sequence" : "Paris is the center of France."
  },
  {
    "token" : "Capital",
    "score" : 0.003259749401778711,
    "sequence" : "Paris is the Capital of France."
  },
  {
    "token" : "centre",
    "score" : 0.002157122475609145,
    "sequence" : "Paris is the centre of France."
  },
  {
    "token" : "city",
    "score" : 9.026127599384262E-4,
    "sequence" : "Paris is the city of France."
  }
]
```

### NER Example
```
POST _ml/trained_models/deployment/bert-model-fine-tuned-for-ner/_infer
{
  "input": "Today's GAH is live from Amsterdam, BC, London, Munich and Texas"
}
```

Returns:
```
[
  {
    "label" : "organisation",
    "score" : 0.940775243737086,
    "word" : "GAH"
  },
  {
    "label" : "location",
    "score" : 0.9987588832004948,
    "word" : "Amsterdam"
  },
  {
    "label" : "location",
    "score" : 0.9958452874139202,
    "word" : "BC"
  },
  {
    "label" : "location",
    "score" : 0.9981461858828271,
    "word" : "London"
  },
  {
    "label" : "location",
    "score" : 0.9991212183928049,
    "word" : "Munich"
  },
  {
    "label" : "location",
    "score" : 0.9994121461792658,
    "word" : "Texas"
  }
]
```

Feature branch PR

Co-authored-by: Dimitris Athanasiou <dimitris@elastic.co>

### CODE REVIEW

**Comment by benwtrent** [comment]
run elasticsearch-ci/bwc
run elasticsearch-ci/default-distro
run elasticsearch-ci/1
run elasticsearch-ci/2
run elasticsearch-ci/docbldesx
run elasticsearch-ci/oss-distro-docs
run elasticsearch-ci/packaging-sample

**Comment by benwtrent** [comment]
run elasticsearch-ci/part-1

**Comment by mark-vieira** [comment]
jenkins test this please

File: x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/inference/nlp/TaskTypeTests.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.inference.nlp;
+
+import org.elasticsearch.test.ESTestCase;
+
+
+public class TaskTypeTests extends ESTestCase {
```
Comment (dimitris-athanasiou) [inline x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/inference/nlp/TaskTypeTests.java:13]: This one is left empty. Should we add some tests here?

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/FillMaskProcessor.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import org.elasticsearch.xpack.core.ml.inference.results.FillMaskResults;
+import org.elasticsearch.xpack.core.ml.inference.results.InferenceResults;
+import org.elasticsearch.xpack.ml.inference.nlp.tokenizers.BertTokenizer;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class FillMaskProcessor implements NlpTask.Processor {
```
Comment (dimitris-athanasiou) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/FillMaskProcessor.java:20]: We should add some tests for this one

**Review by dimitris-athanasiou — COMMENTED** [review]
Looks good. Just a couple of test related comments.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/results/FillMaskResults.java [hunk 3]
```diff
@@ -26,25 +26,25 @@
     public static final String NAME = "fill_mask_result";
     public static final String DEFAULT_RESULTS_FIELD = "results";
```
Comment (dimitris-athanasiou) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/results/FillMaskResults.java:27]: Should this also be `predictions`?

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/FillMaskProcessor.java [hunk 4]
```diff
@@ -27,40 +26,49 @@
         this.bertRequestBuilder = new BertRequestBuilder(tokenizer);
     }
 
+    @Override
+    public void validateInputs(String inputs) {
```
Comment (dimitris-athanasiou) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/FillMaskProcessor.java:30]: nice!

**Review by dimitris-athanasiou — APPROVED** [review]
LGTM Just a question about the name of the fill mask results field. Good to merge though even if you decide to change that.

**Pull request merged**